rfcs/convergence/array-slots.md

```diff
@@ -0,0 +1,165 @@
+RFC: Array slots
+
+---
+
+@smhigley
+
+## Summary
+
+This RFC proposes modifying the Slots API to accept arrays of slot props to support components like Combobox or List, where the most ergonomic authoring experience is passing in an array of data rather than unique JSX children.
+
+## Problem Statement
+
+So far, all slot usage in vNext components assumes the desired authoring experience is to define slots or children one at a time. That makes sense for the components we've created so far, since none of them render a set of elements that are usually generated from an array of data. Even for Menu, which contains a list of MenuItem children, it's not common for authors to want to convert an array of repetitive data into menu items.
+
+In contrast, Combobox and Listbox patterns do commonly take an array of data and use it to render mostly uniform items. The same is likely going to be true of components like List (if we have one) and DetailsList/Grid. For those controls, good authoring ergonomics lean more towards this:
+
+```js
+<Combobox options={['apple', 'banana', 'cucumber', ...etc]} />
+```
+
+Than this:
+
+```js
+<Combobox>
+  <Option>apple</Option>
+  <Option>banana</Option>
+  <Option>cluster fig</Option>
+  [...etc]
+</Combobox>
+```
+
+Accepting an array of slot data isntead of children will also make it easier to internally handle virtualization in the future.
+
+## Detailed Proposal
+
+Modify `getSlots` and `resolveShorthand` to allow authors to pass in an array of slot data for certain slots.
+
+Taking Combobox as an example, the authoring experience would look like this:
+
+```js
+// using strings to define the slots:
+<Combobox options={['apple', 'banana', 'chokecherry']} icon={<Chevron />} ..etc />
+
+// using slot props
+<Combobox options={optionData.map((option, i) => ({ 'data-index': i, children: option.name, title: 'tooltips are awful' }))} icon={<Chevron />} ..etc />
+```
+
+The Combobox component would then handle `options` like this:
+
+within `renderCombobox`:
+
+```js
+const { slots, slotProps } = getSlots < ComboboxSlots > (state, ['options']);
+return (
+  <slots.root {...slotProps.root}>
+    {slots.options.map((Option, i) => {
+      return <Option key={`option-${i}`} {...slotProps.options[i]} />;
```

Comment:
Using Index as a key is an anti-pattern in React:

> We don’t recommend using indexes for keys if the order of items may change. This can negatively impact performance and may cause issues with component state. Check out Robin Pokorny’s article for an in-depth explanation on the negative impacts of using an index as a key. If you choose not to assign an explicit key to list items then React will default to using indexes as keys.

https://reactjs.org/docs/lists-and-keys.html#keys
https://robinpokorny.medium.com/index-as-a-key-is-an-anti-pattern-e0349aece318